mlflow/autogen/autogen_logger.py

```diff
@@ -0,0 +1,262 @@
+            parent_id=self._chat_state.session_span.span_id,
+            name=name,
+            span_type=span_type,
+            inputs=inputs,
+            attributes=attributes,
+            start_time_ns=start_time_ns,
+        )
+
+    def log_event(self, source: Union[str, Agent], name: str, **kwargs: Dict[str, Any]):
+        event_end_time = time.time_ns()
+        if name == "received_message":
+            if (self._chat_state.last_message is not None) and (
+                kwargs.get("sender") not in _EXCLUDED_MESSAGE_SENDERS
+            ):
+                span = self._start_span_in_session(
+                    name=kwargs["sender"],
+                    # Last message is recorded as the input of the next message
+                    inputs=self._chat_state.last_message,
+                    span_type=SpanType.AGENT,
+                    start_time_ns=self._chat_state.last_message_timestamp,
+                )
+                self._client.end_span(
+                    request_id=span.request_id,
+                    span_id=span.span_id,
+                    outputs=kwargs,
+                    end_time_ns=event_end_time,
+                )
+
+                # Re-locate the pended spans under this message span
+                for child_span in self._chat_state.pended_spans:
+                    child_span._span._parent = span._span.context
+                self._chat_state.pended_spans = []
+
+            self._chat_state.last_message = kwargs
+            self._chat_state.last_message_timestamp = event_end_time
+
+    def log_chat_completion(
+        self,
+        invocation_id: uuid.UUID,
+        client_id: int,
+        wrapper_id: int,
+        source: Union[str, Agent],
+        request: Dict[str, Union[float, str, List[Dict[str, str]]]],
+        response: Union[str, ChatCompletion],
+        is_cached: int,
+        cost: float,
+        start_time: str,
+    ) -> None:
+        start_time_epoch = datetime.strptime(start_time, "%Y-%m-%d %H:%M:%S.%f").timestamp()
+        start_time_epoch_ns = int(start_time_epoch * 1e9)
+        end_time = time.time_ns()
+        span = self._start_span_in_session(
+            name="chat_completion",
+            span_type=SpanType.LLM,
+            inputs=request,
+            attributes={
+                "source": source,
+                "client_id": client_id,
+                "invocation_id": invocation_id,
+                "wrapper_id": wrapper_id,
+                "cost": cost,
+                "is_cached": is_cached,
+            },
+            start_time_ns=start_time_epoch_ns,
+        )
+        self._client.end_span(
+            request_id=span.request_id, span_id=span.span_id, outputs=response, end_time_ns=end_time
+        )
+        self._chat_state.pended_spans.append(span)
+
+    # The following methods are not used but are required to implement the BaseLogger interface.
+    def log_function_use(self, *args: Any, **kwargs: Any):
+        pass
```

Comment:
This one seems like it could potentially be quite informative / useful to include in a trace

Reply:
It seems as though a function / tool call is registered the same as any Conversable Agent where you specify `code_excecution_config` in the constructor for the Agent definition to define the tool interface. https://microsoft.github.io/autogen/docs/topics/code-execution/cli-code-executor
I think the general agent span handling will cover it (it will be named whatever user defined agent name is declared "python_tool_agent" "bash_agent" or similar). 
Do you feel that we need to disambiguate these calls separately to make it more clear in the UI, since this will be mapped as a generic "agent" step?

Reply:
The information passed to this method is a subset of what current implementation extracts by wrapping the function call. Using this handler is clearer in terms of readability, but it misses the start time information, which is critical to span creation. Until it supports passing the start time like chat_completion handler, we don't have a good reason to use this handler.

> Do you feel that we need to disambiguate these calls separately to make it more clear in the UI? 

@BenWilson2 Could you elaborate on this? The function call is already rendered as TOOL type in current implementation.

Reply:
I think that the span type identifier is probably completely fine. Provided that the data structure is consistent and we're able to capture the attributes fully, I have zero issues with the current patching.